nservicebus/messaging/transactions.md

```diff
@@ -7,94 +7,106 @@ tags:
+Depending on transport the transaction is escalated to a distributed one (following two-phase commit protocol) when required. For example is not required when using SQL Server transport with NHibernate persistence both targeted at the same database. Since in this case ADO.NET driver guarantees that everything happens inside a single database transaction, ACID guarantees are held for the whole processing.
 
-WARNING: In version 5 and below, when transactions are disabled, no retries will be performed and messages will not be forwarded to the error queue in the event of any failure.
+NOTE: MSMQ will escalate to a distributed transaction right away since it doesn't support promotable transaction enlistments.
 
-snippet:TransactionsDisable
+Ambient transaction mode is enabled by default for the transports that support it. It can be enabled explicitly via
 
+snippet:TransactionsEnable
+
+#### Consistency guarantees
+In this mode handlers will execute inside of the `TransactionScope` created by the transport. This means that all the data updates are executed as a whole or rolled back as a whole.
+
+Should you not want this behavior just select one of the lower transaction modes supported by the selected transport.
 
 ### Transport transaction
 
-In this mode, if supported by the transport, the receive operation is wrapped in a native transaction. The same transaction is used when processing of the message results in sending out other messages. This mode guarantees that the message is not permanently deleted from the incoming queue until at least one processing attempt (including storing user data and sending out messages) is finished successfully.
+In this mode the receive operation is wrapped in a transport native transaction. This mode guarantees that the message is not permanently deleted from the incoming queue until at least one processing attempt (including storing user data and sending out messages) is finished successfully. See the [errors](/nservicebus/errors) section for more details on retries.
 
-NOTE: In this mode messages on the wire can get duplicated at each endpoint so the handler logic has to be designed to be idempotent.
+Use the following code to request *Transport transaction* mode to be used:
 
 snippet:TransactionsDisableDistributedTransactions
 
+#### Multi queue transactions
+Some transports supports enlisting outgoing queue operations in the current receive transaction. This removes the risk of duplicate and ghost messages, see below, being emitted but you still need to make sure that your handlers are idempotent. Currently only the MSMQ and SQL Server transports support multi queue transactions.
```

Comment:
@seanfeldman should I add that note or is it to early to mention it?

On Wed, Dec 9, 2015 at 7:38 PM, Sean Feldman <notifications@github.com>
wrote:

> In nservicebus/messaging/transactions.md
> <https://github.com/Particular/docs.particular.net/pull/971#discussion_r47131856>
> :
>
> >
> >  snippet:TransactionsDisableDistributedTransactions
> >
> > +#### Multi queue transactions
> > +Some transports supports enlisting outgoing queue operations in the current receive transaction. This removes the risk of duplicate and ghost messages, see below, being emitted but you still need to make sure that your handlers are idempotent. Currently only the MSMQ and SQL Server transports support multi queue transactions.
>
> To some extent, the next version of ASB transport (standard or higher
> tier) with a single namespace can support multi queue transactions as well.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/docs.particular.net/pull/971/files#r47131856>
> .
>